lib/spack/spack/ci/generators/snakemake.py

```diff
@@ -0,0 +1,210 @@
+        output_file_path = os.path.abspath(output_file)
+        gen_ci_dir = os.path.dirname(output_file_path)
+        if not os.path.exists(gen_ci_dir):
+            os.makedirs(gen_ci_dir)
+    snakemake_context_dir = os.path.dirname(output_file)
+    if artifacts_root.startswith(snakemake_context_dir):
+        artifacts_root = os.path.relpath(artifacts_root, snakemake_context_dir)
+    pipeline_artifacts_dir = os.path.join(snakemake_context_dir, artifacts_root)
+
+    concrete_env_dir = os.path.join(pipeline_artifacts_dir, "concrete_environment")
+
+    # Now that we've added the mirrors we know about, they should be properly
+    # reflected in the environment manifest file, so copy that into the
+    # concrete environment directory, along with the spack.lock file.
+    if not os.path.exists(concrete_env_dir):
+        os.makedirs(concrete_env_dir)
+    shutil.copyfile(options.env.manifest_path, os.path.join(concrete_env_dir, "spack.yaml"))
+    shutil.copyfile(options.env.lock_path, os.path.join(concrete_env_dir, "spack.lock"))
+
+    update_env_scopes(
+        options.env,
+        [
+            os.path.relpath(s.path, concrete_env_dir)
+            for s in cfg.scopes().values()
+            if isinstance(s, cfg.ImmutableConfigScope) and os.path.exists(s.path)
+        ],
+        os.path.join(concrete_env_dir, "spack.yaml"),
+    )
+
+    job_log_dir = os.path.join(pipeline_artifacts_dir, "logs")
+    job_repro_dir = os.path.join(pipeline_artifacts_dir, "reproduction")
+    job_test_dir = os.path.join(pipeline_artifacts_dir, "tests")
+
+    rel_concrete_env_dir = os.path.relpath(concrete_env_dir, snakemake_context_dir)
+    rel_job_log_dir = os.path.relpath(job_log_dir, snakemake_context_dir)
+    rel_job_repro_dir = os.path.relpath(job_repro_dir, snakemake_context_dir)
+    rel_job_test_dir = os.path.relpath(job_test_dir, snakemake_context_dir)
+
+    spack_version = spack.main.get_version()
+    version_to_clone = spack.main.get_spack_commit() or f"v{spack.spack_version}"
+    rebuild_everything = not options.prune_up_to_date and not options.prune_untouched
+
+    env_vars = {
+        "SPACK_ARTIFACTS_ROOT": artifacts_root,
+        "SPACK_CONCRETE_ENV_DIR": rel_concrete_env_dir,
+        "SPACK_VERSION": spack_version,
+        "SPACK_CHECKOUT_VERSION": version_to_clone,
+        "SPACK_JOB_LOG_DIR": rel_job_log_dir,
+        "SPACK_JOB_REPRO_DIR": rel_job_repro_dir,
+        "SPACK_JOB_TEST_DIR": rel_job_test_dir,
+        "SPACK_PIPELINE_TYPE": options.pipeline_type.name if options.pipeline_type else "None",
+        "SPACK_CI_STACK_NAME": os.environ.get("SPACK_CI_STACK_NAME", "None"),
+        "SPACK_REBUILD_CHECK_UP_TO_DATE": str(options.prune_up_to_date),
+        "SPACK_REBUILD_EVERYTHING": str(rebuild_everything),
+        "SPACK_REQUIRE_SIGNING": str(options.require_signing),
+    }
+
+    spack_ci_ir = spack_ci.generate_ir()
+    rules: List[SnakeRule] = []
+
+    for level, (spec_label, node) in pipeline.traverse():
```

Comment:
This is just a nit, since we don't use the `spec_label`, which iirc is `package/<7 of hash>`, we could just index jobs by dag hash which we do use. And then we can create the `spec_label` if we want it for some reason pretty easily.